File: plantvision/backend/src/routes/equipment.js
const express = require('express');
const { query, validationResult } = require('express-validator');
const db = require('../database/connection');
const { authenticateToken } = require('../middleware/auth');
const auditLogger = require('../utils/auditLogger');
const logger = require('../utils/logger');

const router = express.Router();

// @route   GET /api/v1/equipment
// @desc    Get equipment list with filtering and pagination
// @access  Private
router.get('/', authenticateToken, [
  query('page').optional().isInt({ min: 1 }).withMessage('Page must be a positive integer'),
  query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1 and 100'),
  query('plantId').optional().isUUID().withMessage('Plant ID must be a valid UUID'),
  query('equipmentType').optional().isLength({ min: 1 }).withMessage('Equipment type cannot be empty'),
  query('status').optional().isIn(['active', 'maintenance', 'inactive']).withMessage('Invalid status'),
  query('search').optional().isLength({ min: 1 }).withMessage('Search term cannot be empty')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array()
      });
    }

    const {
      page = 1,
      limit = 50,
      plantId,
      equipmentType,
      status,
      search
    } = req.query;

    const offset = (page - 1) * limit;

    let query = db('equipment')
      .select(
        'equipment.*',
        'plants.plant_code',
        'plants.plant_name'
      )
      .join('plants', 'equipment.plant_id', 'plants.id')
      .orderBy('equipment.equipment_code');

    // Role-based filtering
    if (req.user.role === 'worker' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    }

    // Apply filters
    if (plantId) {
      query = query.where('equipment.plant_id', plantId);
    }

    if (equipmentType) {
      query = query.where('equipment.equipment_type', equipmentType);
    }

    if (status) {
      query = query.where('equipment.status', status);
    }

    if (search) {
      query = query.where(function() {
        this.where('equipment.equipment_code', 'ilike', `%${search}%`)
            .orWhere('equipment.equipment_name', 'ilike', `%${search}%`)
            .orWhere('equipment.description', 'ilike', `%${search}%`);
      });
    }

    // Get total count for pagination
    const totalQuery = query.clone();
    const [{ count: total }] = await totalQuery.count('equipment.id as count');

    // Get paginated results
    const equipment = await query.limit(limit).offset(offset);

    res.json({
      success: true,
      data: {
        equipment: equipment.map(item => ({
          ...item,
          specifications: item.specifications ? JSON.parse(item.specifications) : {}
        })),
        pagination: {
          page: parseInt(page),
          limit: parseInt(limit),
          total: parseInt(total),
          pages: Math.ceil(total / limit)
        }
      }
    });
  } catch (error) {
    logger.error('Get equipment error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment'
    });
  }
});

// @route   GET /api/v1/equipment/:id
// @desc    Get single equipment by ID
// @access  Private
router.get('/:id', authenticateToken, async (req, res) => {
  try {
    const { id } = req.params;

    let query = db('equipment')
      .select(
        'equipment.*',
        'plants.plant_code',
        'plants.plant_name',
        'plants.location as plant_location'
      )
      .join('plants', 'equipment.plant_id', 'plants.id')
      .where('equipment.id', id);

    // Role-based access control
    if (req.user.role === 'worker' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    }

    const equipment = await query.first();

    if (!equipment) {
      return res.status(404).json({
        success: false,
        error: 'Equipment not found'
      });
    }

    // Get recent photos count
    const [{ count: photoCount }] = await db('photos')
      .where({ equipment_id: id })
      .where('status', '!=', 'deleted')
      .count('* as count');

    // Get recent tasks count
    const [{ count: taskCount }] = await db('tasks')
      .where({ equipment_id: id })
      .count('* as count');

    // Log equipment view
    await auditLogger.log({
      userId: req.user.id,
      action: 'VIEW',
      resourceType: 'equipment',
      resourceId: id,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent')
    });

    res.json({
      success: true,
      data: {
        ...equipment,
        specifications: equipment.specifications ? JSON.parse(equipment.specifications) : {},
        photoCount: parseInt(photoCount),
        taskCount: parseInt(taskCount)
      }
    });
  } catch (error) {
    logger.error('Get equipment error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment'
    });
  }
});

// @route   GET /api/v1/equipment/qr/:qrCode
// @desc    Get equipment by QR code
// @access  Private
router.get('/qr/:qrCode', authenticateToken, async (req, res) => {
  try {
    const { qrCode } = req.params;

    let query = db('equipment')
      .select(
        'equipment.*',
        'plants.plant_code',
        'plants.plant_name'
      )
      .join('plants', 'equipment.plant_id', 'plants.id')
      .where('equipment.qr_code', qrCode);

    // Role-based access control
    if (req.user.role === 'worker' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      query = query.where('equipment.location_area', req.user.plantArea);
    }

    const equipment = await query.first();

    if (!equipment) {
      return res.status(404).json({
        success: false,
        error: 'Equipment not found for this QR code'
      });
    }

    // Log QR scan
    await auditLogger.log({
      userId: req.user.id,
      action: 'QR_SCAN',
      resourceType: 'equipment',
      resourceId: equipment.id,
      metadata: { qrCode },
      ipAddress: req.ip,
      userAgent: req.get('User-Agent')
    });

    res.json({
      success: true,
      data: {
        ...equipment,
        specifications: equipment.specifications ? JSON.parse(equipment.specifications) : {}
      }
    });
  } catch (error) {
    logger.error('Get equipment by QR error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment'
    });
  }
});

// @route   GET /api/v1/equipment/types
// @desc    Get unique equipment types
// @access  Private
router.get('/types', authenticateToken, async (req, res) => {
  try {
    let query = db('equipment')
      .distinct('equipment_type')
      .whereNotNull('equipment_type')
      .orderBy('equipment_type');

    // Role-based filtering
    if (req.user.role === 'worker' && req.user.plantArea) {
      query = query.where('location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      query = query.where('location_area', req.user.plantArea);
    }

    const types = await query;

    res.json({
      success: true,
      data: types.map(t => t.equipment_type).filter(Boolean)
    });
  } catch (error) {
    logger.error('Get equipment types error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment types'
    });
  }
});

// @route   GET /api/v1/equipment/:id/photos
// @desc    Get photos for specific equipment
// @access  Private
router.get('/:id/photos', authenticateToken, [
  query('page').optional().isInt({ min: 1 }).withMessage('Page must be a positive integer'),
  query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1 and 100'),
  query('status').optional().isIn(['pending', 'approved', 'rejected']).withMessage('Invalid status'),
  query('startDate').optional().isISO8601().withMessage('Start date must be a valid ISO date'),
  query('endDate').optional().isISO8601().withMessage('End date must be a valid ISO date')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array()
      });
    }

    const { id } = req.params;
    const {
      page = 1,
      limit = 20,
      status,
      startDate,
      endDate
    } = req.query;

    const offset = (page - 1) * limit;

    // Verify equipment exists and user has access
    let equipmentQuery = db('equipment')
      .where('id', id);

    if (req.user.role === 'worker' && req.user.plantArea) {
      equipmentQuery = equipmentQuery.where('location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      equipmentQuery = equipmentQuery.where('location_area', req.user.plantArea);
    }

    const equipment = await equipmentQuery.first();

    if (!equipment) {
      return res.status(404).json({
        success: false,
        error: 'Equipment not found'
      });
    }

    let query = db('photos')
      .select(
        'photos.*',
        'users.username',
        'users.first_name',
        'users.last_name'
      )
      .join('users', 'photos.user_id', 'users.id')
      .where('photos.equipment_id', id)
      .where('photos.status', '!=', 'deleted')
      .orderBy('photos.captured_at', 'desc');

    // Role-based filtering
    if (req.user.role === 'worker') {
      query = query.where('photos.user_id', req.user.id);
    }

    // Apply filters
    if (status) {
      query = query.where('photos.status', status);
    }

    if (startDate) {
      query = query.where('photos.captured_at', '>=', new Date(startDate));
    }

    if (endDate) {
      query = query.where('photos.captured_at', '<=', new Date(endDate));
    }

    // Get total count for pagination
    const totalQuery = query.clone();
    const [{ count: total }] = await totalQuery.count('photos.id as count');

    // Get paginated results
    const photos = await query.limit(limit).offset(offset);

    res.json({
      success: true,
      data: {
        equipment: {
          id: equipment.id,
          equipmentCode: equipment.equipment_code,
          equipmentName: equipment.equipment_name,
          equipmentType: equipment.equipment_type
        },
        photos,
        pagination: {
          page: parseInt(page),
          limit: parseInt(limit),
          total: parseInt(total),
          pages: Math.ceil(total / limit)
        }
      }
    });
  } catch (error) {
    logger.error('Get equipment photos error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment photos'
    });
  }
});

// @route   GET /api/v1/equipment/:id/tasks
// @desc    Get tasks for specific equipment
// @access  Private
router.get('/:id/tasks', authenticateToken, [
  query('page').optional().isInt({ min: 1 }).withMessage('Page must be a positive integer'),
  query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1 and 100'),
  query('status').optional().isIn(['pending', 'in_progress', 'completed', 'cancelled']).withMessage('Invalid status')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        error: 'Validation failed',
        details: errors.array()
      });
    }

    const { id } = req.params;
    const {
      page = 1,
      limit = 20,
      status
    } = req.query;

    const offset = (page - 1) * limit;

    // Verify equipment exists and user has access
    let equipmentQuery = db('equipment')
      .where('id', id);

    if (req.user.role === 'worker' && req.user.plantArea) {
      equipmentQuery = equipmentQuery.where('location_area', req.user.plantArea);
    } else if (req.user.role === 'manager' && req.user.plantArea) {
      equipmentQuery = equipmentQuery.where('location_area', req.user.plantArea);
    }

    const equipment = await equipmentQuery.first();

    if (!equipment) {
      return res.status(404).json({
        success: false,
        error: 'Equipment not found'
      });
    }

    let query = db('tasks')
      .select(
        'tasks.*',
        'assigned_user.username as assigned_username',
        'assigned_user.first_name as assigned_first_name',
        'assigned_user.last_name as assigned_last_name',
        'assigner.username as assigner_username',
        'assigner.first_name as assigner_first_name',
        'assigner.last_name as assigner_last_name'
      )
      .join('users as assigned_user', 'tasks.assigned_to', 'assigned_user.id')
      .join('users as assigner', 'tasks.assigned_by', 'assigner.id')
      .where('tasks.equipment_id', id)
      .orderBy('tasks.created_at', 'desc');

    // Role-based filtering
    if (req.user.role === 'worker') {
      query = query.where('tasks.assigned_to', req.user.id);
    }

    // Apply filters
    if (status) {
      query = query.where('tasks.status', status);
    }

    // Get total count for pagination
    const totalQuery = query.clone();
    const [{ count: total }] = await totalQuery.count('tasks.id as count');

    // Get paginated results
    const tasks = await query.limit(limit).offset(offset);

    res.json({
      success: true,
      data: {
        equipment: {
          id: equipment.id,
          equipmentCode: equipment.equipment_code,
          equipmentName: equipment.equipment_name,
          equipmentType: equipment.equipment_type
        },
        tasks,
        pagination: {
          page: parseInt(page),
          limit: parseInt(limit),
          total: parseInt(total),
          pages: Math.ceil(total / limit)
        }
      }
    });
  } catch (error) {
    logger.error('Get equipment tasks error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to retrieve equipment tasks'
    });
  }
});

module.exports = router;

File: plantvision/mobile/lib/features/equipment/domain/entities/equipment.dart
class Equipment {
  final String id;
  final String plantId;
  final String equipmentCode;
  final String equipmentName;
  final String? equipmentType;
  final String? description;
  final String? manufacturer;
  final String? model;
  final String? serialNumber;
  final DateTime? installationDate;
  final String? locationArea;
  final double? latitude;
  final double? longitude;
  final String? qrCode;
  final Map<String, dynamic> specifications;
  final EquipmentStatus status;
  final String? plantCode;
  final String? plantName;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Equipment({
    required this.id,
    required this.plantId,
    required this.equipmentCode,
    required this.equipmentName,
    this.equipmentType,
    this.description,
    this.manufacturer,
    this.model,
    this.serialNumber,
    this.installationDate,
    this.locationArea,
    this.latitude,
    this.longitude,
    this.qrCode,
    this.specifications = const {},
    required this.status,
    this.plantCode,
    this.plantName,
    required this.createdAt,
    required this.updatedAt,
  });

  String get displayName => '$equipmentCode - $equipmentName';
  
  String get fullLocation {
    final parts = <String>[];
    if (plantName != null) parts.add(plantName!);
    if (locationArea != null) parts.add(locationArea!);
    return parts.join(' - ');
  }

  bool get hasLocation => latitude != null && longitude != null;

  Equipment copyWith({
    String? id,
    String? plantId,
    String? equipmentCode,
    String? equipmentName,
    String? equipmentType,
    String? description,
    String? manufacturer,
    String? model,
    String? serialNumber,
    DateTime? installationDate,
    String? locationArea,
    double? latitude,
    double? longitude,
    String? qrCode,
    Map<String, dynamic>? specifications,
    EquipmentStatus? status,
    String? plantCode,
    String? plantName,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Equipment(
      id: id ?? this.id,
      plantId: plantId ?? this.plantId,
      equipmentCode: equipmentCode ?? this.equipmentCode,
      equipmentName: equipmentName ?? this.equipmentName,
      equipmentType: equipmentType ?? this.equipmentType,
      description: description ?? this.description,
      manufacturer: manufacturer ?? this.manufacturer,
      model: model ?? this.model,
      serialNumber: serialNumber ?? this.serialNumber,
      installationDate: installationDate ?? this.installationDate,
      locationArea: locationArea ?? this.locationArea,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      qrCode: qrCode ?? this.qrCode,
      specifications: specifications ?? this.specifications,
      status: status ?? this.status,
      plantCode: plantCode ?? this.plantCode,
      plantName: plantName ?? this.plantName,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'plantId': plantId,
      'equipmentCode': equipmentCode,
      'equipmentName': equipmentName,
      'equipmentType': equipmentType,
      'description': description,
      'manufacturer': manufacturer,
      'model': model,
      'serialNumber': serialNumber,
      'installationDate': installationDate?.millisecondsSinceEpoch,
      'locationArea': locationArea,
      'latitude': latitude,
      'longitude': longitude,
      'qrCode': qrCode,
      'specifications': specifications,
      'status': status.name,
      'plantCode': plantCode,
      'plantName': plantName,
      'createdAt': createdAt.millisecondsSinceEpoch,
      'updatedAt': updatedAt.millisecondsSinceEpoch,
    };
  }

  factory Equipment.fromJson(Map<String, dynamic> json) {
    return Equipment(
      id: json['id'] as String,
      plantId: json['plant_id'] ?? json['plantId'] as String,
      equipmentCode: json['equipment_code'] ?? json['equipmentCode'] as String,
      equipmentName: json['equipment_name'] ?? json['equipmentName'] as String,
      equipmentType: json['equipment_type'] ?? json['equipmentType'] as String?,
      description: json['description'] as String?,
      manufacturer: json['manufacturer'] as String?,
      model: json['model'] as String?,
      serialNumber: json['serial_number'] ?? json['serialNumber'] as String?,
      installationDate: json['installation_date'] != null || json['installationDate'] != null
          ? DateTime.parse((json['installation_date'] ?? json['installationDate']) as String)
          : null,
      locationArea: json['location_area'] ?? json['locationArea'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      qrCode: json['qr_code'] ?? json['qrCode'] as String?,
      specifications: json['specifications'] as Map<String, dynamic>? ?? {},
      status: EquipmentStatus.fromString(json['status'] as String),
      plantCode: json['plant_code'] ?? json['plantCode'] as String?,
      plantName: json['plant_name'] ?? json['plantName'] as String?,
      createdAt: DateTime.parse((json['created_at'] ?? json['createdAt']) as String),
      updatedAt: DateTime.parse((json['updated_at'] ?? json['updatedAt']) as String),
    );
  }
}

enum EquipmentStatus {
  active,
  maintenance,
  inactive;

  static EquipmentStatus fromString(String status) {
    switch (status.toLowerCase()) {
      case 'active':
        return EquipmentStatus.active;
      case 'maintenance':
        return EquipmentStatus.maintenance;
      case 'inactive':
        return EquipmentStatus.inactive;
      default:
        throw ArgumentError('Invalid equipment status: $status');
    }
  }

  String get displayName {
    switch (this) {
      case EquipmentStatus.active:
        return 'Active';
      case EquipmentStatus.maintenance:
        return 'Under Maintenance';
      case EquipmentStatus.inactive:
        return 'Inactive';
    }
  }

  bool get isOperational => this == EquipmentStatus.active;
}